### ISSUE TITLE
[systemd-networkd] virtual machine's network is sometimes not brought up correctly

### ISSUE THREAD

**Issue by jelly** [systemd/systemd#22237]
**systemd version the issue has been seen with**

250

**Used distribution**

Arch Linux

**Linux kernel version used** (`uname -a`)

Linux archlinux 5.16.2-arch1-1 #1 SMP PREEMPT Thu, 20 Jan 2022 16:18:29 +0000 x86_64 GNU/Linux

**CPU architecture issue was seen on**

x86_64 (virtualized)

**Expected behaviour you didn't see**

A simple dhcp setup stopped working with the upgrade to 250 from 249.7-2. The issue does not always occur but as [cockpits tests](https://github.com/cockpit-project/bots/pull/2796) reboot regularly it often occurs and the tests have to be retried. This was not the case with systemd 249. The virtual machine uses qemu's networking.

```
-device virtio-rng-pci,rng=objrng0,id=rng0,bus=pci.0,addr=0x6 -netdev socket,mcast=230.0.0.1:5500,id=mcast0 -device virtio-net-pci,netdev=mcast0,mac=52:54:01:00:00:01,bus=pci.0,addr=0x0f -netdev user,id=base0,restrict=on,net=172.27.0.0/24,dnssearch=loopback,hostname=arch-127-0-0-2-2201,hostfwd=tcp:127.0.0.2:2201-:22,hostfwd=tcp:127.0.0.2:9091-:9090 -device virtio-net-pci,netdev=base0,bus=pci.0,addr=0x0e -sandbox on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny -msg timestamp=on
```

networkd setup:
```
[Match]
Name=eth0

[Network]
IPv6LinkLocalAddressGenerationMode=none
DHCP=yes
```

Note that `IPv6LinkLocalAddressGenerationMode=none` was added later to check if networkd was unsuccesfully waiting on a DHCPv6 lease

**Unexpected behaviour you saw**

Sometimes after a reboot systemd-networkd is stuck

**Steps to reproduce the problem**
 
That's kinda tricky, I can provide the virtual machine if required.

**Additional program output to the terminal or log subsystem illustrating the issue**

I've collected systemd-networkd logs with debug loglevel, this includes the unsuccessfully first and then systemd-networkd is restarted and it all works fine. 
```text
Jan 24 11:49:46 archlinux systemd-networkd[260]: Enumeration completed
Jan 24 11:49:46 archlinux systemd[1]: Started Network Configuration.
Jan 24 11:49:46 archlinux systemd-networkd[260]: Bus bus-api-network: changing state AUTHENTICATING → HELLO
Jan 24 11:49:46 archlinux systemd-networkd[260]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Jan 24 11:49:46 archlinux systemd-networkd[260]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RequestName cookie=2 reply_cookie=0 signature=su error-name=n/a error-message=n/a
...skipping...
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: link_check_ready(): link layer is configuring.
Jan 24 11:51:03 archlinux systemd-networkd[421]: Bus bus-api-network: changing state OPENING → AUTHENTICATING
Jan 24 11:51:03 archlinux systemd-networkd[421]: Registering bus object implementation for path=/org/freedesktop/network1 iface=org.freedesktop.network1.Manager
Jan 24 11:51:03 archlinux systemd-networkd[421]: Registering bus object implementation for path=/org/freedesktop/network1/link iface=org.freedesktop.network1.DHCPServer
Jan 24 11:51:03 archlinux systemd-networkd[421]: Registering bus object implementation for path=/org/freedesktop/network1/link iface=org.freedesktop.network1.Link
Jan 24 11:51:03 archlinux systemd-networkd[421]: Registering bus object implementation for path=/org/freedesktop/network1/network iface=org.freedesktop.network1.Network
Jan 24 11:51:03 archlinux systemd-networkd[421]: Registering bus object implementation for path=/org/freedesktop/LogControl1 iface=org.freedesktop.LogControl1
Jan 24 11:51:03 archlinux systemd-networkd[421]: sd-device-monitor: Failed to stat PID1's netns, ignoring: No such file or directory
Jan 24 11:51:03 archlinux systemd-networkd[421]: Loaded timestamp for '/etc/systemd/network'.
Jan 24 11:51:03 archlinux systemd-networkd[421]: Loaded timestamp for '/run/systemd/network'.
Jan 24 11:51:03 archlinux systemd-networkd[421]: /etc/systemd/network/20-wired.network: DHCPv6 client is enabled but IPv6 link local addressing is disabled. Disabling DHCPv6 client.
Jan 24 11:51:03 archlinux systemd-networkd[421]: Virtualization QEMU found in DMI (/sys/class/dmi/id/sys_vendor)
Jan 24 11:51:03 archlinux systemd-networkd[421]: UML virtualization not found in /proc/cpuinfo.
Jan 24 11:51:03 archlinux systemd-networkd[421]: Virtualization XEN not found, /proc/xen does not exist
Jan 24 11:51:03 archlinux systemd-networkd[421]: Virtualization found, CPUID=KVMKVMKVM
Jan 24 11:51:03 archlinux systemd-networkd[421]: Found VM virtualization kvm
Jan 24 11:51:03 archlinux systemd-networkd[421]: /usr/lib/systemd/network/80-container-host0.network: Conditions in the file do not match the system environment, skipping.
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: Saved new link: ifindex=3, iftype=ETHER(1), kind=n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: Saved permanent hardware address: 52:54:01:00:00:01
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: Saved original MTU 1500 (min: 68, max: 65535)
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: Saved hardware address: 52:54:01:00:00:01
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: Link UP
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: Gained carrier
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: udev initialized link
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: State changed: pending -> initialized
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: Saved new link: ifindex=2, iftype=ETHER(1), kind=n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: Saved permanent hardware address: 52:54:00:12:34:56
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: Saved original MTU 1500 (min: 68, max: 65535)
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: Saved hardware address: 52:54:00:12:34:56
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: Link UP
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: Gained carrier
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: udev initialized link
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: State changed: pending -> initialized
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Saved new link: ifindex=1, iftype=LOOPBACK(772), kind=n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Saved permanent hardware address: 00:00:00:00:00:00
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Failed to get driver, continuing without: Operation not supported
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Saved original MTU 65536 (min: 0, max: 4294967295)
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Saved hardware address: 00:00:00:00:00:00
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Flags change: +LOOPBACK +UP +LOWER_UP +RUNNING
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Link UP
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Gained carrier
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: udev initialized link
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: State changed: pending -> initialized
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Received new foreign address (configured): ::1/128 (valid forever, preferred forever), flags: permanent, scope: host
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: link_check_ready(): link is in initialized state.
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Received new foreign address (configured): 127.0.0.1/8 (valid forever, preferred forever), flags: permanent, scope: host
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: link_check_ready(): link is in initialized state.
Jan 24 11:51:03 archlinux systemd-networkd[421]: rtnl: received non-static neighbor, ignoring.
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth1: Received new foreign route (configured): dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: multicast, nexthop: 0, priority: 256, flags: n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: eth0: Received new foreign route (configured): dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: multicast, nexthop: 0, priority: 256, flags: n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Received new foreign route (configured): dst: ::1/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: local, nexthop: 0, priority: 0, flags: n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Received new foreign route (configured): dst: ::1/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main(254), proto: kernel, type: unicast, nexthop: 0, priority: 256, flags: n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Received new foreign route (configured): dst: 127.255.255.255/32, src: n/a, gw: n/a, prefsrc: 127.0.0.1, scope: link, table: local(255), proto: kernel, type: broadcast, nexthop: 0, priority: 0, flags: n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Received new foreign route (configured): dst: 127.0.0.1/32, src: n/a, gw: n/a, prefsrc: 127.0.0.1, scope: host, table: local(255), proto: kernel, type: local, nexthop: 0, priority: 0, flags: n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: lo: Received new foreign route (configured): dst: 127.0.0.0/8, src: n/a, gw: n/a, prefsrc: 127.0.0.1, scope: host, table: local(255), proto: kernel, type: local, nexthop: 0, priority: 0, flags: n/a
Jan 24 11:51:03 archlinux systemd-networkd[421]: Remembering foreign routing policy rule (configured): priority: 32766, ::/0 -> ::/0, iif: n/a, oif: n/a, table: main(254)
Jan 24 11:51:03 archlinux systemd-networkd[421]: Remembering foreign routing policy rule (configured): priority: 0, ::/0 -> ::/0, iif: n/a, oif: n/a, table: local(255)
Jan 24 11:51:03 archlinux systemd-networkd[421]: rtnl: received rule message with invalid family 129, ignoring.
Jan 24 11:51:03 archlinux systemd-networkd[421]: rtnl: received rule message with invalid family 128, ignoring.
Jan 24 11:51:03 archlinux systemd-networkd[421]: Remembering foreign routing policy rule (configured): priority: 32767, 0.0.0.0/0 -> 0.0.0.0/0, iif: n/a, oif: n/a, table: default(253)
Jan 24 11:51:03 archlinux systemd-networkd[421]: Remembering foreign routing policy rule (configured): priority: 32766, 0.0.0.0/0 -> 0.0.0.0/0, iif: n/a, oif: n/a, table: main(254)
Jan 24 11:51:03 archlinux systemd-networkd[421]: Remembering foreign routing policy rule (configured): priority: 0, 0.0.0.0/0 -> 0.0.0.0/0, iif: n/a, oif: n/a, table: local(255)
Jan 24 11:51:03 archlinux systemd-networkd[421]: Enumeration completed
Jan 24 11:51:03 archlinux systemd[1]: Started Network Configuration.
Jan 24 11:51:03 archlinux systemd-networkd[421]: Bus bus-api-network: changing state AUTHENTICATING → HELLO
Jan 24 11:51:03 archlinux systemd-networkd[421]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
```

As it's a VM and I have to restart systemd-networkd to get working ssh access, a screenshot of `ip a`
![image](https://user-images.githubusercontent.com/67428/150778736-192b2a56-7fd8-4202-8ee4-1e86010d6c96.png)

Without `IPv6LinkLocalAddressGenerationMode=none`
![image](https://user-images.githubusercontent.com/67428/150788523-c19f2de0-e294-42b6-9a6d-5400001380c3.png)

**Comment by yuwata** [comment]
Could you provide full debugging logs of networkd?

**Comment by jelly** [comment]
[systemd-networkd.log](https://github.com/systemd/systemd/files/7926631/systemd-networkd.log)

**Comment by yuwata** [comment]
Thanks. I will take a look.

**Comment by yuwata** [comment]
Hm, that's quite strange. Why the DHCPv4 address is once configured but soon removed? If it is removed by networkd, then there should exist logs `Removing address: ...`, but there is no logs about that. I guess that another network interface management daemon is running on the system.

**Comment by yuwata** [comment]
OK. Let's close this. If you still find any issues, then please feel free to request to reopen this issue, or open a new issue page. Thank you.
